edlib/src/edlib.tpp

```diff
@@ -1417,51 +1440,46 @@ static int obtainAlignmentHirschberg(
  * @return  Alphabet as a string of unique characters, where index of each character is its value in transformed
  *          sequences.
  */
-static string transformSequences(const char* const queryOriginal, const int queryLength,
-                                 const char* const targetOriginal, const int targetLength,
-                                 unsigned char** const queryTransformed,
-                                 unsigned char** const targetTransformed) {
+template <class Element, class AlphabetIdx>
+static unordered_map<Element, AlphabetIdx> edlib::internal::transformSequences(const Element* const queryOriginal, const int queryLength,
+                                                              const Element* const targetOriginal, const int targetLength,
+                                                              AlphabetIdx** const queryTransformed,
+                                                              AlphabetIdx** const targetTransformed) {
     // Alphabet is constructed from letters that are present in sequences.
     // Each letter is assigned an ordinal number, starting from 0 up to alphabetLength - 1,
     // and new query and target are created in which letters are replaced with their ordinal numbers.
     // This query and target are used in all the calculations later.
-    *queryTransformed = static_cast<unsigned char *>(malloc(sizeof(unsigned char) * queryLength));
-    *targetTransformed = static_cast<unsigned char *>(malloc(sizeof(unsigned char) * targetLength));
+    *queryTransformed = static_cast<AlphabetIdx *>(malloc(sizeof(AlphabetIdx) * queryLength));
+    *targetTransformed = static_cast<AlphabetIdx *>(malloc(sizeof(AlphabetIdx) * targetLength));
 
-    string alphabet = "";
+    unordered_map<Element, AlphabetIdx> elementToAlphabetIdx;
 
     // Alphabet information, it is constructed on fly while transforming sequences.
-    // letterIdx[c] is index of letter c in alphabet.
-    unsigned char letterIdx[MAX_UCHAR + 1];
-    bool inAlphabet[MAX_UCHAR + 1]; // inAlphabet[c] is true if c is in alphabet
-    for (int i = 0; i < MAX_UCHAR + 1; i++) inAlphabet[i] = false;
-
+    // elementToAlphabetIdx[c] is index of letter c in alphabet.
+    AlphabetIdx currentSize =0 ;
     for (int i = 0; i < queryLength; i++) {
-        unsigned char c = static_cast<unsigned char>(queryOriginal[i]);
-        if (!inAlphabet[c]) {
-            inAlphabet[c] = true;
-            letterIdx[c] = static_cast<unsigned char>(alphabet.size());
-            alphabet += queryOriginal[i];
+        Element c = queryOriginal[i];
+        if (elementToAlphabetIdx.find(c) == elementToAlphabetIdx.end()) {
```

Comment:
My question from before: should we do .find(c) just once and then use that, instead of doing it three times?

Reply:
You're right. We can call `find(c)` just once and store its output in an iterator and use the iterator multiple times.
We can also store the iterator of the end in another iterator and update it just when an insertion is happening. By doing that there is no need to call `end()` for each if statement.

Reply:
Storing find(c) in iterator -> yes, let's do that.
Return value of .end() -> I think we can leave that as it is, I would expect unordered_map to internally track the end() iterator, I doubt it recalculates it each time.

Reply:
I just googled, cpp std lib guarantees constant complexity for getting the end() of container.